### PR TITLE
View to Component

### PR DESCRIPTION
**Pull request by Archmonger**
- #41 

## Changelog
- `view_to_component` utility that supports both async and sync views. Views can be function or classes.
   - Can convert view to VDOM or an iframe (for compatibility purposes)
- Reorganize the docs to have separate `src` an `includes` folders 
- Docs for `view_to_component`
- Update changelog
- Miscellaneous cleanup across several files

## IMPORTANT NOTE

This PR does not add automatic handling of things such as href click events or form submission. Currently `compatibility=True` is required to get href and forms to work.

This is a desirable behavior that should be implemented in a future PR as a toggleable option.

In the future `view_to_component` should be able to automatically perform any URL routing/re-rendering that is needed. The VDOM component should essentially "act as if it's an iframe", despite it being VDOM.

## Remaining Todo:
- [x] Tests (needs merge of Playwright test suite)
- [x] Docs
- [x] Type Hints

### CODE REVIEW

**Comment by Archmonger** [comment]
Blocked on:
- https://github.com/idom-team/idom/issues/777

**Comment by Archmonger** [comment]
@rmorshea I'm having some issues with `@idom.component` type hints.

Between these two files
- `tests/test_app/components.py`
- `src/django_idom/components.py`

Seems like immediately returning a component within a component causes type hint issues. Additionally, embedding a `view_to_component` within VDOM such as `idom.html.div` can cause type hint issues as well.

**Comment by Archmonger** [comment]
@rmorshea Just merged some `mypy` warning fixes, but that's really a kick in the teeth for my `pylance` dev environment.

`mypy` does not like `type: ignore` in situations where it doesn't believe it needs it. And `pylance` has several more type warnings than `mypy` that are detected.

Can we consider using [`pyright`](https://github.com/microsoft/pyright) (the type checker built into `pylance`) instead of `mypy`?

**Comment by Archmonger** [comment]
@rmorshea Just merged with `main`, so this PR is ready for review. There's some [features](https://github.com/idom-team/django-idom/pull/81#issue-1288122062) I'd like to add to `view_to_component` in the future, but I don't believe that should be addressed in this PR.

I will make a new release as soon as this gets merged in.

File: src/django_idom/components.py
Comment (rmorshea) [inline src/django_idom/components.py:52]: I'm also realizing that `lake8_idom_hooks` does not show errors for [hook usages after early returns](https://github.com/idom-team/flake8-idom-hooks/issues/15).

**Review by rmorshea — CHANGES_REQUESTED** [review]

File: tests/test_app/templates/base.html
Comment (Archmonger) [reply tests/test_app/templates/base.html:46]: What's the purpose of a newline following HTML?

Seems like nearly all formatters I have seen strips the last newline, without the ability to configure this behavior.

File: src/django_idom/components.py
Comment (rmorshea) [reply src/django_idom/components.py:55]: The problem with `args` and `kwargs` is that it's hard to tell if their values change.

Comment (Archmonger) [reply src/django_idom/components.py:55]: This is actually a bug I'm realizing. I would like the preserve the ability to re-render if `args` or `kwargs` change.

In the future, we'll also want to trigger a re-render when HTTP methods change (which is why it originally was outside of the `async_renderer`. 

For now, I will set `dependencies=[args, kwargs]`

Comment (rmorshea) [reply src/django_idom/components.py:55]: So the pattern here would be to do the following:

```python
def view_to_component(view, ..., dependencies):
    view = use_callback(view, dependencies=dependencies)
    rendered_view, set_rendered_view = use_state(None)

    @use_effect
    async def render_view():
        set_rendered_view(view())

    return rendered_view
```

Usage being:

```python
view_to_component(view)
view_to_component(lambda: view(...))
view_to_component(lambda: view(...), dependencies=...)
```

Comment (Archmonger) [reply src/django_idom/components.py:55]: > The problem with `args` and `kwargs` is that it's hard to tell if their values change.

Do you have a proposed solution for this?

Comment (rmorshea) [reply src/django_idom/components.py:55]: `dependencies=[args, kwargs]` won't quite work either since the user is probably doing:

```
view_to_component(view, args=(...), kwargs={...})
```

And the identity of `args` and `kwargs` will change every time. You could expand all the values within `args` and `kwargs`, but that get's pretty hacky.

The pattern I suggest above, of requiring the user to wrap functions that require more arguments, is pretty common in React. It has the benefit of allowing us to only check the values we know will change by leveraging the magic of `use_callback`.

Comment (Archmonger) [reply src/django_idom/components.py:55]: Wrapping the view with `lambda` completely murders the current rendering schema. Definitely going to bring significantly inefficiency surrounding async views.

Comment (rmorshea) [reply src/django_idom/components.py:55]: Is there any way that the logic could be re-worked to check `inspect.isawaitable` instead of `iscoroutinefunction`?

If not, perhaps you could restrict parameters to positional arguments? That's at least a little easier to destructure and pass to `dependencies` in the effect:

```python
def view_to_component(view, *args, ...):
    
    @use_effect(dependencies=[view, *args])
    async def render_view(): ...
```

Comment (rmorshea) [reply src/django_idom/components.py:55]: Actually, `isawaitable` doesn't seem viable. I'd go with the second option. Limit parameters to positional args.

Comment (rmorshea) [reply src/django_idom/components.py:55]: Or, if you really want to allow kwargs you'll have to do something like the following:

```python
dependencies = [..., *args] + [x for item in kwargs.items() for x in item]
```

Comment (Archmonger) [reply src/django_idom/components.py:55]: Unfortunately, kwargs are technically the more important thing here.

In theory, views should always be passed serializable data since the values [come from either URLs parts](https://docs.djangoproject.com/en/4.1/topics/http/urls/#example) or URL parameters. So maybe we can do `dependencies = [request, *args, json.dumps(kwargs)]`?

Comment (rmorshea) [reply src/django_idom/components.py:55]: Ah, that's a clever solution. You'll probably need to silence some errors from `flake8_idom_hooks` but this should be fine.

Comment (rmorshea) [reply src/django_idom/components.py:55]: Err, if you want to `json.dumps(kwargs)` you should do the same for `args` to keep the behavior consistent. De-structuring `args` into dependencies will make the effect check the identity of each value for changes. Whereas, with `json.dumps` effectively causes the effect to check for value changes.

Comment (Archmonger) [reply src/django_idom/components.py:55]: I'm thinking of doing the same for the request object, to avoid mutability.

Something similar to `json.dumps(vars(request_obj))`

Perhaps with `default=lambda x: None` to avoid callables from breaking things.

Comment (Archmonger) [reply src/django_idom/components.py:66]: I think the latest solution is the best we're going to get.

On a different note, `flake8-idom-hooks` identifies `rendered_view, set_rendered_view = hooks.use_state(None)` and `@hooks.use_effect` as `hook 'use_effect' used outside component or hook definition`

Comment (rmorshea) [reply src/django_idom/components.py]: I haven't released the latest `flake8-idom-hooks` yet, so it's not considering the `@component` decorator.

Comment (Archmonger) [reply src/django_idom/components.py]: I changed the component name to CamelCase for testing. Looks like it doesn't give any warnings, so no need to suppress anything.

Is the latest implementation good to go?

Comment (rmorshea) [reply src/django_idom/components.py]: Are there any tests that check whether the component will re-render appropriately if inputs change?

Comment (Archmonger) [reply src/django_idom/components.py]: I will make one today

Comment (Archmonger) [reply src/django_idom/components.py]: New tests are live

Comment (rmorshea) [reply src/django_idom/components.py]: Also just released the latest flake8 plugin.

Comment (rmorshea) [reply src/django_idom/components.py]: Will review the new tests on Sunday

File: tests/test_app/components.py
Comment (rmorshea) [inline tests/test_app/components.py:331]: Instead of depending on these 3 second delays, can we do something a little more deterministic and add a button that can be clicked to trigger these changes? Not necessary, but if you go with this button approach you could probably combine `view_to_component_request`, `view_to_component_args`, and `view_to_component_kwargs` together into one component with separate buttons to trigger each.

**Review by rmorshea — CHANGES_REQUESTED** [review]

File: tests/test_app/components.py
Comment (Archmonger) [reply tests/test_app/components.py:331]: We aren't depending on the 3 second delays. I just added them as a "human readable" way to tell if the test is working.

Should I completely remove the delay?

Also I think separate views is more appropriate. Pytest styling conventions say all tests should be written out long-form and not combined (Always `WET`, never `DRY`). It helps improve readability and traceability of exception stacks.

**Comment by Archmonger** [comment]
@rmorshea I'm realizing a limitation with IDOM Core's `view_to_component`.

Currently, because string based event handlers such as `<button onclick='example()'></button>` aren't allowed in core, `onclick` events cannot be serialized into VDOM.

**Comment by rmorshea** [comment]
This is actually more of an issue with React. If you assign `onclick="..."` IDOM will pass that on to the client as-is. However React does not interpret strings as event handlers. I think the only way to do this would be to [`dangerouslySetInnerHtml`](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) at which point it doesn't seem like you'd need `html_to_vdom` at all since you could literally use the raw HTML string.

**Comment by Archmonger** [comment]
I would say this is ends up being a pretty big limitation. When working within JavaScript frameworks, this issue doesn't really exist. This is really a problem specific to us, due to the cross-language barrier.

I can't think of a safe way of fixing this.

_**cough**_ Even our existing `script` tag implementation is potentially unsafe.

Might want to document the limitation in core, probably within the `html_to_vdom` docstring.

**Comment by Archmonger** [comment]
On a different note, I think the revised tests for script, request, args, and kwargs are solid.

**Review by rmorshea — APPROVED** [review]

**Pull request merged**
